expr/func_aspercent.go

```diff
@@ -0,0 +1,285 @@
+		totalsSerie.QueryPatt = fmt.Sprint(s.totalFloat)
+	}
+
+	for _, serie := range series {
+		serie.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie.QueryPatt, totalsSerie.QueryPatt)
+		serie.Target = fmt.Sprintf("asPercent(%s,%s)", serie.Target, totalsSerie.QueryPatt)
+		serie.Tags = map[string]string{"name": serie.Target}
+		for i := range serie.Datapoints {
+			var totalVal float64
+			if len(totalsSerie.Datapoints) > i {
+				totalVal = totalsSerie.Datapoints[i].Val
+			} else {
+				totalVal = s.totalFloat
+			}
+			serie.Datapoints[i].Val = computeAsPercent(serie.Datapoints[i].Val, totalVal)
+		}
+		outSeries = append(outSeries, serie)
+		cache[Req{}] = append(cache[Req{}], serie)
+	}
+
+	return outSeries, nil
+}
+
+func computeAsPercent(in, total float64) float64 {
+	if math.IsNaN(in) || math.IsNaN(total) {
+		return math.NaN()
+	}
+	if total == 0 {
+		return math.NaN()
+	}
+	return in / total * 100
+}
+
+func groupSeriesByKey(series []models.Series, nodes []expr, keys *[]string) map[string][]models.Series {
+	keyedSeries := make(map[string][]models.Series)
+Loop:
+	for _, serie := range series {
+		key := aggKey(serie, nodes)
+		if _, ok := keyedSeries[key]; !ok {
+			keyedSeries[key] = []models.Series{serie}
+			for _, k := range *keys {
+				if k == key {
+					continue Loop
+				}
+			}
+			*keys = append(*keys, key)
+		} else {
+			keyedSeries[key] = append(keyedSeries[key], serie)
+		}
+	}
+	return keyedSeries
+}
+
+// Sums each seriesList in map of seriesLists
+func getTotalSeries(totalSeriesLists map[string][]models.Series) map[string]models.Series {
+	totalSeries := make(map[string]models.Series, len(totalSeriesLists))
+	for key := range totalSeriesLists {
+		totalSeries[key] = sumSeries(totalSeriesLists[key])
+	}
+	return totalSeries
+}
+
+// Sums seriesList
+func sumSeries(series []models.Series) models.Series {
+	var summedSerie models.Series
+	if len(series) == 1 {
+		summedSerie = series[0]
+	} else {
+		out := pointSlicePool.Get().([]schema.Point)
+		crossSeriesSum(series, &out)
+		var queryPatts []string
+
+	Loop:
+		for _, v := range series {
+			// avoid duplicates
+			for _, qp := range queryPatts {
+				if qp == v.QueryPatt {
+					continue Loop
+				}
+			}
+			queryPatts = append(queryPatts, v.QueryPatt)
+		}
+		name := fmt.Sprintf("sumSeries(%s)", strings.Join(queryPatts, ","))
+		cons, queryCons := summarizeCons(series)
+		summedSerie = models.Series{
+			Target:       name,
+			QueryPatt:    name,
+			Datapoints:   out,
+			Interval:     series[0].Interval,
+			Consolidator: cons,
+			QueryCons:    queryCons,
+			Tags:         map[string]string{"name": name},
+		}
+	}
+	return summedSerie
+}
+
+func aggKey(serie models.Series, nodes []expr) string {
+	metric := extractMetric(serie.Target)
+	if len(metric) == 0 {
+		metric = serie.Tags["name"]
+	}
+	// Trim off tags (if they are there) and split on '.'
+	parts := strings.Split(strings.SplitN(metric, ";", 2)[0], ".")
+	var name []string
+	for _, n := range nodes {
```

Comment:
Looks similar to aliasByTags. Might be worth factoring this out (groupByNodes implementation would likely use this as well)?

Reply:
moved function to parse.go and updated aliasByNode to use the function